protocols/kad/src/behaviour.rs

```diff
@@ -981,31 +981,44 @@ where
         // overridden as it avoids having to load the existing record in the
         // first place.
 
-        // The record is cloned because of the weird libp2p protocol requirement
-        // to send back the value in the response, although this is a waste of
-        // resources.
-        match self.store.put(record.clone()) {
-            Ok(()) => {
-                debug!("Record stored: {:?}; {} bytes", record.key, record.value.len());
-                self.queued_events.push_back(NetworkBehaviourAction::NotifyHandler {
-                    peer_id: source,
-                    handler: NotifyHandler::One(connection),
-                    event: KademliaHandlerIn::PutRecordRes {
-                        key: record.key,
-                        value: record.value,
-                        request_id,
-                    },
-                })
-            }
-            Err(e) => {
-                info!("Record not stored: {:?}", e);
-                self.queued_events.push_back(NetworkBehaviourAction::NotifyHandler {
-                    peer_id: source,
-                    handler: NotifyHandler::One(connection),
-                    event: KademliaHandlerIn::Reset(request_id)
-                })
+        if !record.is_expired(Instant::now()) {
+            // The record is cloned because of the weird libp2p protocol
+            // requirement to send back the value in the response, although this
+            // is a waste of resources.
+            match self.store.put(record.clone()) {
+                Ok(()) => debug!("Record stored: {:?}; {} bytes", record.key, record.value.len()),
+                Err(e) => {
+                    info!("Record not stored: {:?}", e);
+                    self.queued_events.push_back(NetworkBehaviourAction::NotifyHandler {
+                        peer_id: source,
+                        handler: NotifyHandler::One(connection),
+                        event: KademliaHandlerIn::Reset(request_id)
+                    });
+
+                    return
```

Comment:
I would personally prefer to put the remaining code in this branch into an `else` clause for `if !record.is_expired(now)`, because I find early `return`s in the middle of larger code blocks / functions to often trip me up in reasoning about the control flow (as compared to early returns at the very beginning of a function).

Reply:
But then I would need to duplicate the happy case return logic. We want to return `PutRecordRes` both if:

1. The record is expired.

2. The record is successfully put into the record store.

I would end up with:

```diff
diff --git a/protocols/kad/src/behaviour.rs b/protocols/kad/src/behaviour.rs
index 0e6035ad..b34580af 100644
--- a/protocols/kad/src/behaviour.rs
+++ b/protocols/kad/src/behaviour.rs
@@ -986,7 +986,19 @@ where
             // requirement to send back the value in the response, although this
             // is a waste of resources.
             match self.store.put(record.clone()) {
-                Ok(()) => debug!("Record stored: {:?}; {} bytes", record.key, record.value.len()),
+                Ok(()) => {
+                    debug!("Record stored: {:?}; {} bytes", record.key, record.value.len());
+
+                    self.queued_events.push_back(NetworkBehaviourAction::NotifyHandler {
+                        peer_id: source,
+                        handler: NotifyHandler::One(connection),
+                        event: KademliaHandlerIn::PutRecordRes {
+                            key: record.key,
+                            value: record.value,
+                            request_id,
+                        },
+                    })
+                },
                 Err(e) => {
                     info!("Record not stored: {:?}", e);
                     self.queued_events.push_back(NetworkBehaviourAction::NotifyHandler {
@@ -994,31 +1006,29 @@ where
                         handler: NotifyHandler::One(connection),
                         event: KademliaHandlerIn::Reset(request_id)
                     });
-
-                    return
                 }
             }
+        } else {
+            // The remote receives a [`KademliaHandlerIn::PutRecordRes`] even in the
+            // case where the record is discarded due to being expired. Given that
+            // the remote sent the local node a [`KademliaHandlerEvent::PutRecord`]
+            // request, the remote perceives the local node as one node among the k
+            // closest nodes to the target. Returning a [`KademliaHandlerIn::Reset`]
+            // instead of an [`KademliaHandlerIn::PutRecordRes`] to have the remote
+            // try another node would only result in the remote node to contact an
+            // even more distant node. In addition returning
+            // [`KademliaHandlerIn::PutRecordRes`] does not reveal any internal
+            // information to a possibly malicious remote node.
+            self.queued_events.push_back(NetworkBehaviourAction::NotifyHandler {
+                peer_id: source,
+                handler: NotifyHandler::One(connection),
+                event: KademliaHandlerIn::PutRecordRes {
+                    key: record.key,
+                    value: record.value,
+                    request_id,
+                },
+            })
         }
-
-        // The remote receives a [`KademliaHandlerIn::PutRecordRes`] even in the
-        // case where the record is discarded due to being expired. Given that
-        // the remote sent the local node a [`KademliaHandlerEvent::PutRecord`]
-        // request, the remote perceives the local node as one node among the k
-        // closest nodes to the target. Returning a [`KademliaHandlerIn::Reset`]
-        // instead of an [`KademliaHandlerIn::PutRecordRes`] to have the remote
-        // try another node would only result in the remote node to contact an
-        // even more distant node. In addition returning
-        // [`KademliaHandlerIn::PutRecordRes`] does not reveal any internal
-        // information to a possibly malicious remote node.
-        self.queued_events.push_back(NetworkBehaviourAction::NotifyHandler {
-            peer_id: source,
-            handler: NotifyHandler::One(connection),
-            event: KademliaHandlerIn::PutRecordRes {
-                key: record.key,
-                value: record.value,
-                request_id,
-            },
-        })
     }
```

Am I missing something?

Reply:
I overlooked that, apologies.